### PR TITLE
Update supervisor exit reason

### PR DESCRIPTION
**Pull request by FelonEkonom**
Related Jira ticket: https://membraneframework.atlassian.net/browse/MC-203

### CODE REVIEW

**Comment by FelonEkonom** [comment]
This will fix https://github.com/membraneframework/membrane_core/issues/566

File: lib/membrane/core/process.ex [hunk 1]
```diff
@@ -0,0 +1,24 @@
+defmodule Membrane.Core.Process do
```
Comment (mat-hek) [inline lib/membrane/core/process.ex:1]:
```suggestion
defmodule Membrane.Core.ProcessHelper do
```

File: lib/membrane/core/process.ex [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+      _any when not log? -> :ok
+      :normal -> :ok
+      :shutdown -> :ok
+      {:shutdown, _reason} -> :ok
+      _reason -> Membrane.Logger.error("Terminating with reason: #{inspect(reason)}")
+    end
+
+    # this is a hack to exit with a custom reason, but without having GenServer
+    # exit logs occurring when the exit reason is neither :normal, :shutdown
+    # nor {:shutdown, reason}
```
Comment (mat-hek) [inline lib/membrane/core/process.ex:20]: move that to the doc of this function

File: lib/membrane/core/process.ex [hunk 3]
```diff
@@ -0,0 +1,24 @@
+defmodule Membrane.Core.Process do
+  @moduledoc false
+
+  require Membrane.Logger
+
+  @spec exit_self(any(), log?: boolean()) :: no_return()
+  def exit_self(reason, opts \\ []) do
```
Comment (mat-hek) [inline lib/membrane/core/process.ex:7]:
```suggestion
  def take_french_leave(reason, opts \\ []) do
```

File: lib/membrane/core/parent/child_life_controller.ex [hunk 4]
```diff
@@ -692,7 +693,7 @@ defmodule Membrane.Core.Parent.ChildLifeController do
         Terminating.
         """)
 
-        exit({:shutdown, :child_crash})
+        Core.Process.exit_self({:child_crash, child_name})
```
Comment (mat-hek) [inline lib/membrane/core/parent/child_life_controller.ex:696]:
```suggestion
        Core.Process.exit_self({:membrane_child_crash, child_name})
```

Comment (mat-hek) [reply lib/membrane/core/parent/child_life_controller.ex:696]: I'd also not call exit from this function directly, but return `{:stop, reason}` and handle that in the calling module

**Review by mat-hek — CHANGES_REQUESTED** [review]

File: lib/membrane/core/parent/child_life_controller.ex [hunk 5]
```diff
@@ -667,21 +669,24 @@ defmodule Membrane.Core.Parent.ChildLifeController do
...
+        |> case do
+          {:removed, state} ->
+            exec_handle_crash_group_down_callback(
+              group.name,
+              group.members,
+              group.crash_initiator || child_name,
+              state
+            )
+
+          {_result, state} ->
```
Comment (mat-hek) [inline lib/membrane/core/parent/child_life_controller.ex:684]:
```suggestion
          {:not_removed, state} ->
```

File: lib/membrane/core/parent/child_life_controller.ex [hunk 6]
```diff
@@ -640,22 +639,25 @@ defmodule Membrane.Core.Parent.ChildLifeController do
...
 
-    if state.terminating? and Enum.empty?(state.children) do
-      {:stop, state}
-    else
-      {:continue, state}
+      {:ok, state} when state.terminating? and state.children == %{} ->
+        ProcessHelper.notoelo(:normal)
+
+      {:ok, state} ->
+        state
```
Comment (mat-hek) [inline lib/membrane/core/parent/child_life_controller.ex:652]: consider returning `{:continue, state} | {:stop, reason, state}` and handling that in bin.ex/pipeline.ex

**Review by mat-hek — APPROVED** [review]

**Pull request merged**